src/goto-programs/remove_exceptions.cpp

```diff
@@ -0,0 +1,598 @@
+
+  void instrument_exception_handlers(
+    const goto_functionst::function_mapt::iterator &);
+
+  void add_gotos(
+    const goto_functionst::function_mapt::iterator &);
+
+  void add_throw_gotos(
+    const goto_functionst::function_mapt::iterator &,
+    const goto_programt::instructionst::iterator &,
+    const stack_catcht &);
+
+  void add_function_call_gotos(
+    const goto_functionst::function_mapt::iterator &,
+    const goto_programt::instructionst::iterator &,
+    const stack_catcht &);
+};
+
+/*******************************************************************\
+
+Function: remove_exceptionst::add_exceptional_returns
+
+Inputs:
+
+Outputs:
+
+Purpose: adds exceptional return variables for every function that
+         may escape exceptions
+
+\*******************************************************************/
+
+void remove_exceptionst::add_exceptional_returns(
+  const goto_functionst::function_mapt::iterator &func_it)
+{
+  const irep_idt &function_id=func_it->first;
+  goto_programt &goto_program=func_it->second.body;
+
+  assert(symbol_table.has_symbol(id2string(function_id)));
+  symbolt function_symbol=symbol_table.lookup(id2string(function_id));
+
+  // for now only add exceptional returns for Java
+  if(function_symbol.mode!=ID_java)
+    return;
+
+  if(goto_program.empty())
+    return;
+
+  // We generate an exceptional return value for any function that has
+  // a throw or a function call. This can be improved by only considering
+  // function calls that may escape exceptions. However, this will
+  // require multiple passes.
+  Forall_goto_program_instructions(instr_it, goto_program)
+  {
+    if(instr_it->is_throw() || instr_it->is_function_call())
+    {
+      // look up the function symbol
+      symbol_tablet::symbolst::iterator s_it=
+        symbol_table.symbols.find(function_id);
+
+      assert(s_it!=symbol_table.symbols.end());
+      const symbolt &function_symbol=s_it->second;
+
+      auxiliary_symbolt new_symbol;
+      new_symbol.is_static_lifetime=true;
+      new_symbol.module=function_symbol.module;
+      new_symbol.base_name=id2string(function_symbol.base_name)+EXC_SUFFIX;
+      new_symbol.name=id2string(function_symbol.name)+EXC_SUFFIX;
+      new_symbol.mode=function_symbol.mode;
+      new_symbol.type=typet(ID_pointer, empty_typet());
+      symbol_table.add(new_symbol);
+
+      // initialize the exceptional return with NULL
+      symbol_exprt lhs_expr_null;
+      lhs_expr_null.set_identifier(id2string(function_symbol.name)+EXC_SUFFIX);
+      lhs_expr_null.type()=typet(ID_pointer, empty_typet());
+
+      exprt rhs_expr_null;
+      rhs_expr_null=null_pointer_exprt(pointer_typet());
+      goto_programt::targett t_null=goto_program.insert_before(instr_it);
+      t_null->make_assignment();
+      t_null->source_location=instr_it->source_location;
+      t_null->code=code_assignt(
+        lhs_expr_null,
+        typecast_exprt(rhs_expr_null, lhs_expr_null.type()));
```

Comment:
If the subtype is set properly as suggested by my above comment, then you shouldn't even need any typecast here.